packages/gatsby/src/internal-plugins/webpack-theme-component-shadowing/index.js

```diff
@@ -0,0 +1,74 @@
+const path = require(`path`)
+
+module.exports = class GatsbyThemeComponentShadowingResolverPlugin {
+  cache = {}
+
+  constructor({ projectRoot, themes }) {
+    this.themes = themes
+    this.projectRoot = projectRoot
+  }
+
+  apply(resolver) {
+    resolver.plugin(`relative`, (request, callback) => {
+      // find out which theme's src/components dir we're requiring from
+      const matchingThemes = this.themes.filter(name =>
+        request.path.includes(path.join(name, `src`, `components`))
+      )
+      // 0 matching themes happens a lot fo rpaths we don't want to handle
+      // > 1 matching theme means we have a path like
+      //   `gatsby-theme-blog/src/components/gatsby-theme-something/src/components`
+      if (matchingThemes.length > 1) {
```

Comment:
as an inital comment, the core theming algorithm doesn't recurse yet, so you can't construct child-of-child themes yet. (This is intended and I need to PR for that functionality).

---

This brings up another good point in that we'll want a CLI option to list "overridable components". 

---

To answer your initial question, the path

```
gatsby-theme-saas-marketing/src/components/gatsby-theme-blog/BlogPost.js
```

only includes a single match:

```
gatsby-theme-saas-marketing/src/components
```

So the shadowing position from a user's perspective would be

```
my-site/src/components/gatsby-theme-saas-marketing/gatsby-theme-blog/BlogPost.js
```

---

This *is* a potentially interesting issue of 

```
import BlogPost from 'gatsby-theme-blog/src/components/BlogPost.js
```

Which for now, I've decided to not include in shadowing since it's an explicit import of a third party component and since the user has control over the import, they can adjust it to be the new component instead of having an overridden magical import that points to something that isn't actually imported.

Reply:
> the core theming algorithm doesn't recurse yet, so you can't construct child-of-child themes yet. (This is intended and I need to PR for that functionality).

Cool. I'll sit tight on that line of thinking until you say so.

> This brings up another good point in that we'll want a CLI option to list "overridable components".

In theory, wouldn't this be `ls {theme}/src/components`?

> This is a potentially interesting issue of
> 
> ```
> import BlogPost from 'gatsby-theme-blog/src/components/BlogPost.js
> ```

A good question here would be how much should be automatic vs. opt-in.

For example, in WordPress theming is automatic: _any_ file in a theme can be overridden by adding a file by the same name in a child theme. (This has limitations, I believe, and it's been a long time since I did WP dev, so forgive me if I’m wrong here.)

The benefit to this approach is that it imposes fewer demands on theme authors to strictly define their APIs. The downside, of course, is that many themes are a goddamn free-for-all because there are hardly any requirements.

On the other hand, if we have some kind of explicit opt-in, it means that we can control things in a way that improves the DX (e.g. auto-mapping overridable components to `import BlogPost from '{theme}/blog-post'`). The trade-off is that theme authors have to do more upfront work to create that DX. Plus whatever extra code we have to write to make the opt-in API work.

I'm not sure any of this is relevant to this PR, but just in case there are larger API implications here I want to bring it up now.

Reply:
> In theory, wouldn't this be ls {theme}/src/components?

More or less, but it'd be nice to provide a list that does the shadowing logic for you and shows you were to put things maybe?

> A good question here would be how much should be automatic vs. opt-in.

For now, I've opted to make src/components the only automatic shadowing directory. It is possible that we open this up into more locations before themes ship as a stable feature. I think that to start, being opt-in by creating files in src/components is the path of least resistance for theme authors that also allows some level of control (for example, should the gatsby-* files be shadowable? probably not?).

Also, as I talk about a bit in my other comment on this PR, I see a progressive disclosure to theming that onboards people into creating their own themes as they try to take control over more and more of the site (answering the question, how much of a theme can you override/modify/shadow before its no longer useful to be using someone else's theme?). Shadowing is an early-stage step in that process, moving up into the gatsby lifecycle methods later on.